mozilla-services/socorro
# bug 1433148: switch from boto to boto3

This switches the code from boto to boto3. Because that effectively required rewriting some of the code, I ended up rewriting all of it.
    
The S3Connection class is based on the one we have in Antenna with some changes:
    
* uses configman instead of everett
* has a `load_file` method
* nixes the bucket check code
* has better tests
    
Because `S3Connection` supports the local dev environment and server environments and a variety of different AWS authentication methods, this removes all the other connection context variants since we don't need them anymore.
    
This redoes the tests for crash_data and crashstorage.
    
This adds docstrings for methods along with notes and what the methods might raise.
    
Since `S3Connection` handles retrying, this removes the retrying in the crashstorage code.
    
This changes some of the configuration. We no longer have `AWS_HOST`, `AWS_PORT`, and `AWS_SECURE`. Those have been replaced with `AWS_ENDPOINT_URL`. This only affects the local dev environment.
    
This adds `ResourceWarning` to the warnings to ignore in pytest. boto3 raises a `ResourceWarning` since it doesn't explicitly close connections.

This gets rid of moto in favor of tests hitting localstack-s3 service directly.

Test plan:

1. `make build`, `make setup`, `make lint`, `make test`
2. process some crashes and make sure they show up in webapp
   1. `make run` in one terminal
   2. `make shell` in another and then `socorro-cmd fetch_crashids --num=5 | ./scripts/process_crashes.sh`
   3. wait a bit and then open one of the links
3. run verifyprocessed script: `cd webapp-django && ./manage.py verifyprocessed`
4. run upload_telemetry_schema script:
   1. `socorro-cmd upload_telemetry_schema`
   2. verify it created the `.json` file with `./scripts/socorro_aws_s3.sh ls --recursive s3://telemetry_bucket/`
5. check a crash with `http://localhost:8000/api/CrashVerify/?crash_id=UUID`
6. list the bucket and make sure paths start with `v1/` and `v2/`:
   1. `./scripts/socorro_aws_s3.sh ls --recursive s3://dev_bucket/`
   2. `./scripts/socorro_aws_s3.sh ls --recursive s3://telemetry_bucket/`

I think that's it. It verifies that keys didn't change, that everything works with itself, that the processor can load and save crash data, the telemetry crash storage works, uploading the telemetry schema works, and verifyprocessed works.
